### PR TITLE
[Backend] Use pod namespace as default when not in multi user mode

### PR DESCRIPTION
**Pull request by Bobgy**
Fixes https://github.com/kubeflow/pipelines/issues/2908

The fix is the same as comment: https://github.com/kubeflow/pipelines/issues/2908#issuecomment-582872532

Problem of the bug is that, pod client need to default to pod namespace when it is not provided.

My fix: I think it's better to be explicit about which namespace we want to use, so I added getting namespace logic in handlers and removed default namespace logic in workflow client.
In addition to that, now I did a little more sanity checking depending on multi user mode or not:
1. When in multi user mode, report error when namespace is not provided.
2. Otherwise, use pod namespace as default
3. Workflow client no longer default to pod namespace when provided namespace is empty.
4. ReportWorkflow endpoint is a little different, because it should be called by persistence agent who gets data from argo workflows. I changed it to always expect namespace in the workflow.

Verification: I built an image and patched to kfp 0.2.2 standalone deployment, verified now pods can be retried successfully.

/assign @jingzhang36 
/assign @IronPan

---
This change is [Reviewable](https://reviewable.io/reviews/kubeflow/pipelines/3009)

### CODE REVIEW

File: backend/src/apiserver/common/config.go [hunk 1]
```diff
@@ -23,7 +23,9 @@ import (
 )
 
 const (
-	MultiUserMode     string = "MULTIUSER"
+	MultiUserMode       string = "MULTIUSER"
+	podNamespace        string = "POD_NAMESPACE"
+	podNamespaceDefault string = "kubeflow"
```
Comment (jingzhang36) [inline backend/src/apiserver/common/config.go]: a quick question: the standalone deployment using "kubeflow" as default, the hosted use "kubeflow" or "default"? Does it matter in this case?

Comment (Bobgy) [reply backend/src/apiserver/common/config.go]: It doesn't matter. The default branch will only trigger in tests.

In k8s env, POD_NAMESPACE environment variable is always provided to each pod, so api server will get the correct value injected.

Comment (jingzhang36) [reply backend/src/apiserver/common/config.go]: I see.
/lgtm

Comment (Bobgy) [reply backend/src/apiserver/common/config.go]: Hold on, actually that's not set by k8s, but by us: https://github.com/kubeflow/pipelines/search?q=POD_NAMESPACE&unscoped_q=POD_NAMESPACE

So I guess it's better to report an Error if it's not provided.

Comment (Bobgy) [reply backend/src/apiserver/common/config.go]: I've changed to not default to kubeflow and init tests with POD_NAMESPACE env var

File: backend/src/apiserver/resource/resource_manager.go [hunk 2]
```diff
@@ -599,6 +607,9 @@ func (r *ResourceManager) ReportWorkflowResource(workflow *util.Workflow) error
 	}
 	runId := workflow.ObjectMeta.Labels[util.LabelKeyWorkflowRunId]
 	jobId := workflow.ScheduledWorkflowUUIDAsStringOrEmpty()
+	if len(workflow.Namespace) == 0 {
+		return util.NewInvalidInputError("Workflow missing namespace")
```
Comment (IronPan) [inline backend/src/apiserver/resource/resource_manager.go]: Why is this needed? 
Are we sure this would work for "default" Kubernetes namespace?

Comment (Bobgy) [reply backend/src/apiserver/resource/resource_manager.go]: This is not needed. But I think it is clearer, because I think persistence agent is the only instance calling this. And it gets data using workflow client, so it will always carry namespace.

Comment (Bobgy) [reply backend/src/apiserver/resource/resource_manager.go]: I verified retry works. That includes coverage for persistence agent. Also, e2e tests pass.

**Comment by Bobgy** [comment]
/retest
Cloud build flakiness

**Comment by Bobgy** [comment]
strange, a lot of samples failed
/retest
